cranelift/codegen/src/isa/riscv64/inst/imms.rs

```diff
@@ -0,0 +1,237 @@
+        let bit = 1 << 11;
+        if val == 0 {
+            Some(Imm12 { bits: 0 })
+        } else if (val & bit) != 0 && (val >> 12) == 0xffff_ffff_ffff_f {
+            Some(Imm12 {
+                bits: (val & 0xffff) as i16,
+            })
+        } else if (val & bit) == 0 && (val >> 12) == 0 {
+            Some(Imm12 {
+                bits: (val & 0xffff) as i16,
+            })
+        } else {
+            None
+        }
+    }
+
+    #[inline(always)]
+    pub fn from_bits(bits: i16) -> Self {
+        Self { bits: bits & 0xfff }
+    }
+    #[inline(always)]
+    pub(crate) fn form_bool(b: bool) -> Self {
+        if b {
+            Self { bits: -1 }
+        } else {
+            Self { bits: 0 }
+        }
+    }
+
+    /// Create a zero immediate of this format.
+    #[inline(always)]
+    pub fn zero() -> Self {
+        Imm12 { bits: 0 }
+    }
+    #[inline(always)]
+    pub fn as_i16(self) -> i16 {
+        self.bits
+    }
+    #[inline(always)]
+    pub fn as_u32(&self) -> u32 {
+        (self.bits as u32) & 0xfff
+    }
+}
+
+impl Into<i64> for Imm12 {
+    fn into(self) -> i64 {
+        self.bits as i64
+    }
+}
+impl Display for Imm12 {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
+        write!(f, "{:+}", self.bits)
+    }
+}
+
+impl std::ops::Neg for Imm12 {
+    type Output = Self;
+    fn neg(self) -> Self::Output {
+        Self { bits: -self.bits }
+    }
+}
+
+// singed
+#[derive(Clone, Copy, Default)]
+pub struct Imm20 {
+    /// The immediate bits.
+    pub bits: i32,
+}
+
+impl Imm20 {
+    #[inline(always)]
+    pub fn from_bits(bits: i32) -> Self {
+        Self {
+            bits: bits & 0xf_ffff,
+        }
+    }
+    #[inline(always)]
+    pub fn as_u32(&self) -> u32 {
+        (self.bits as u32) & 0xf_ffff
+    }
+}
+
+impl Debug for Imm20 {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
+        write!(f, "{}", self.bits)
+    }
+}
+
+impl Display for Imm20 {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
+        write!(f, "{}", self.bits)
+    }
+}
+
+#[derive(Clone, Copy)]
+pub struct Uimm5 {
+    bits: u8,
+}
+
+impl Uimm5 {
+    pub fn maybe_from_u64(val: u64) -> Option<Uimm5> {
+        if (val >> 5) == 0 {
+            Some(Self {
+                bits: (val & 0b1_1111) as u8,
+            })
+        } else {
+            None
+        }
+    }
+    pub fn from_bits(bits: u8) -> Self {
+        Self { bits }
+    }
+    /// Create a zero immediate of this format.
+    pub fn zero() -> Self {
+        Self { bits: 0 }
+    }
+    pub fn as_u32(&self) -> u32 {
+        (self.bits as u32) & 0b1_1111
+    }
+}
+
+impl Debug for Uimm5 {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
+        write!(f, "{}", self.bits)
+    }
+}
+
+impl Display for Uimm5 {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
+        write!(f, "{}", self.bits)
+    }
+}
+
+/*
+    a imm20 immediate and a Imm12 immediate can generate what immediate.
+    imm20 + imm12
+*/
```

Comment:
Please use doc comments.

```suggestion
/// An imm20 immediate and a Imm12 immediate can generate what immediate.
/// imm20 + imm12
```

Also I think you should use standalone functions instead of a dummy type.

Reply:
ok. solved.